TM: leverage Levenshtein distance on ES hits

julen opened the pull request:
This leverages the Levenshtein distance on top of the hits returned by ES, considering the whole text, not individual terms.

In my testing the irrelevant results are now gone. As mentioned in the issue, we set a threshold to allow changes up to a third of the text contents. YMMV so feel free to propose any adjustments to the threshold.

The PR also removes the `MIN_SCORE` setting because it results irrelevant (check the commit message for further reasoning).

Fixes #3971.

unho commented:
Shouldn't removal of `MIN_SCORE` be recorded in release notes?

unho commented:
Shouldn't removal of MIN_SCORE be recorded in release notes?

unho commented on pootle/core/search/backends/elasticsearch.py:
```diff
@@ -26,6 +28,26 @@
+
+    Any hits with a normalized distance value (0..1) higher than
+    `max_distance` will be discarded. It's assumed that `hits` is already
+    sorted from higher to lower score.
```
You shouldn't assume that. Even more, [score is altered below in this file by using the TM weight](https://github.com/translate/pootle/blob/master/pootle/core/search/backends/elasticsearch.py#L113).

unho replied:
Leaving that post-processing change it seems that we can rely on [elasticsearch returning the results sorted by relevance](https://www.elastic.co/guide/en/elasticsearch/guide/current/sorting.html).

julen replied:
It's perfectly fine to assume that. Even if you multiply the score by a weight factor later, those relevance scores are still a) sorted, b) refer to irrelevant results.

dwaynebailey commented:
@julen So my reading is we just do ``AUTO`` all the time now and use LS to filter the results.  So there is now no way to actually set a threshold score.  Would it not be better to allow the threshold to be set in the config?

julen commented:
I have no problem to allow setting the threshold value externally. How did testing go with the PR by the way?

julen commented:
I updated the PR with some of the proposed fixes, and the threshold can now be configured with `MAX_DISTANCE`. Looking forward to hear about your test results!

dwaynebailey commented:
@julen ran some tests, and tried to fake some LS distances.  Pretty hard to determine which should match.  But the matching works correctly as far as I can tell. Would be helpful if the ``data-score`` was actually a levenshtein value.  I've got values of 10.NNN for my tests so hard to know if this is even filtering edit distance values of e.g. 0.3

Otherwise lgtm.

julen commented:
Oh, good catch, actually I should remove the `data-score` values from the DOM — they were meant to help debug scores, but we've seen these are irrelevant.

dwaynebailey commented:
@julen might be useful if they actually where LS distances.  Certainly helps quickly confirm that things are working as expected.  So I'd support changing them to LS and keeping them.

julen commented:
I added an extra commit which changes this to LS distances, I'd appreciate if you can double-check now. Still I'd be in favor of not transmitting such _debugging_ information over the wire. We need to add some tests for the backend code definitely, although at the moment I'm pretty confident all results over the threshold are being discarded.

dwaynebailey commented:
@julen retested.  TBH having the data there is helpful so I wouldn't want to remove it.  I ended up adding my own debug statements so that I could see what was accepted and what was rejected.  Please add some if you feel it may be useful, but not critical.

GTM for the code.

My only concern is this.  AUTO in ES was meant to adjust the acceptable distance depending on string length.  While this didn't work I did notice that a fake string of 'One' won't match 'Ones', i.e. for really short strings we won't get matches that we might want.  I'm not 100% sure its worth worrying about though.

julen commented:
I updated the PR taking into account @friedelwolff's suggestion to use similarity rather than distance for consistency with the rest of our tools. I also added debug information via logging so anyone can check what's going on behind the scenes. I consider this is good to go from my side.

dwaynebailey commented:
lgtm.  Didn't test, but was happy with my last batch of testing.
